client/blocks/image-selector/preview.jsx

```diff
@@ -0,0 +1,236 @@
+/** @format */
+
+/**
+ * External dependencies
+ */
+
+import PropTypes from 'prop-types';
+import React, { Component } from 'react';
+import { defer, isEqual, uniq } from 'lodash';
+import classNames from 'classnames';
+import Gridicon from 'gridicons';
+import { localize } from 'i18n-calypso';
+
+/**
+ * Internal dependencies
+ */
+import Button from 'components/button';
+import ImagePreloader from 'components/image-preloader';
+import MediaActions from 'lib/media/actions';
+import MediaStore from 'lib/media/store';
+import Spinner from 'components/spinner';
+import { url } from 'lib/media/utils';
+
+export class ImageSelectorPreview extends Component {
+	static propTypes = {
+		siteId: PropTypes.number,
+		itemIds: PropTypes.array,
+		multiple: PropTypes.bool,
+		maxWidth: PropTypes.number,
+		onImageChange: PropTypes.func,
+		showEditIcon: PropTypes.bool,
+		translate: PropTypes.func,
+	};
+
+	state = {
+		images: [],
+		transientImages: {},
+	};
+
+	componentDidMount() {
+		this.fetchImages();
+		MediaStore.on( 'change', this.updateImageState );
+	}
+
+	componentDidUpdate( prevProps ) {
+		const { siteId, itemIds } = this.props;
+		if ( siteId !== prevProps.siteId || itemIds !== prevProps.itemIds ) {
+			this.fetchImages();
+		}
+	}
+
+	componentWillUnmount() {
+		MediaStore.off( 'change', this.updateImageState );
+	}
+
+	setTransientImages = () => {
+		const newTransientImages = {};
+		this.state.images.forEach( image => {
+			if ( image.transient ) {
+				const media = MediaStore.get( this.props.siteId, image.ID );
+				newTransientImages[ media.ID ] = image.URL;
+			}
+		} );
+		this.setState( { transientImages: { ...this.state.transientImages, ...newTransientImages } } );
+	};
+
+	fetchImages = () => {
+		// We may not necessarily need to trigger a network request if we
+		// already have the data for the media item, so first update the state
+		this.updateImageState( () => {
+			const { itemIds, siteId } = this.props;
+			if ( isEqual( this.state.images.map( image => image.ID ), itemIds ) ) {
+				return;
+			}
+
+			defer( () => {
```

Comment:
Why do we need the `defer` here?

Reply:
I believe this and the other defer you mentioned below are from the initial commit before image fetching was moved to the setState callback.  

They were needed to make sure the images were fetched correctly on mount, but I'm unable to reproduce right now so I think they can be safely removed.

Reply:
👍